onnx/defs/rnn/defs.cc

```diff
@@ -5,124 +5,311 @@
-          "bidirectional",
-          AttrType::INT)
-    .Attr("skip_input_transform",
-          "If set, skips linear transformation on the input of the first layer",
-          AttrType::INT)
-    .Attr("num_layers", "Numbers of RNN layers in the stack, default 1",
-          AttrType::INT)
+  - Ct = ft (.) Ct-1 + it (.) ct
+  - H = ot (.) tanh(Ct)
+)DOC")
+    .Attr("activations", "A list of 4 activation functions for input, output, "
+	  "forget, and cell gates. Typical activation functions are sigmoid "
+	  "and tanh. See the equations for default if not specified",
+          AttrType::STRINGS)
     .Attr("hidden_size", "Number of neurons in the hidden layer", AttrType::INT)
-    .Input(0, "weights", R"DOC(
-All parameters of the stack packed together in the opaque tensor. The size must
-be compatible with input attributes passed to the op.
-
-The layout format is the one used by CuDNN and very similar to TensorRT:
-
-The weight structure holds weights and biases for each layer of the network.
-Each parameter matrix is linearly appended after the previous parameter matrix
-without padding.
-
-The order of matrixes `{K, L, D, R, N, C}` is defined as:
- - K - type of the matrix: `weight` (first) or `bias` second
- - L - The number of layers in the RNN - `num_layers`
- - D - The direction of the layer: normal (first) or reverse (second).
-                                   (in case of `directions=2`)
- - R - The type of the connection: `input-hidden` (first) or
-                                   `hidden-hidden` (second)
- - N - The number of gates matrices in the RNN, dependent on the `cell_type`:
- -- For `relu` or `tanh` there is one gate
- -- For `gru` there are 3 gates ordered as `reset`, `update`, `hidden`
- -- For `lstm` there are 4 gates ordered as `input`, `forget`, `cell`, `output`
- - C - The size of each matrix, which varies.
- -- If the linear layer on the input is skipped (`skip_input_transform=1`)
-    and then for the first layer (`L=1`) the weight matrix (`K=weight`)
-    on the input connection (`R=input-hidden`) is skipped,
-    i.e. has 0 parameters in the list
- -- For the first layer (`L=1`) weight matrix (`K=weight`) on input connection
-    (`R=input-hidden`), dimensions are `{hidden_size, input_size}`
- -- For other layers (`L>1`) weight matrix (`K=weight`) on input connection
-    (`R=input-hidden`), dimensions are `{hidden_size, directions * hidden_size}`
- -- For weight matrix (`K=weight`) on recurrent connection (`R=hidden-hidden`),
-    dimensions are `{hidden_size, hidden_size}`
- -- For all biases (`K=bias`), dimensions are `{hidden_size}`
+    .Attr("reverse", "Process the sequences in reverse order if 1, default 0.",
+          AttrType::INT)
+    .Attr("input_forget", "Couple the input and forget gates if 1, default 0.",
+          AttrType::INT)	  
+    .Attr("clip", "Cell clip threshold. Clipping bounds the elements of a tensor "
```

Comment:
for repetitive arguments like this you may find FillUsing useful and extract common args to a subfunction. See for example https://github.com/onnx/onnx/blob/ea6d6578d9d2e566be69077a8c5b43663b0649c7/onnx/defs/reduction/defs.cc#L36

Reply:
Good point. Done.